executor/window_parallel.go

```diff
@@ -0,0 +1,367 @@
+			return
+		}
+		numRows := chk.NumRows()
+		for i := 0; i < numRows; i++ {
+			workerIdx := int(murmur3.Sum32(groupKeys[i])) % len(e.workers)
+			w := &e.workers[workerIdx]
+
+			if results[workerIdx] == nil {
+				select {
+				case <-e.finishCh:
+					return
+				case results[workerIdx], ok = <-w.inputHolderCh:
+					if !ok {
+						return
+					}
+				}
+			}
+			results[workerIdx].AppendRow(chk.GetRow(i))
+			if results[workerIdx].IsFull() {
+				w.inputCh <- results[workerIdx]
+				results[workerIdx] = nil
+			}
+		}
+	}
+	for i := range e.workers {
+		if results[i] != nil {
+			e.workers[i].inputCh <- results[i]
+			results[i] = nil
+		}
+	}
+}
+
+func (e *WindowParallelExec) prepare4ParallelExec(ctx context.Context) {
+	go e.execDataFetcherThread(ctx)
+
+	waitGroup := &sync.WaitGroup{}
+	waitGroup.Add(len(e.workers))
+	for i := range e.workers {
+		go e.workers[i].run(ctx, waitGroup)
+	}
+
+	go e.waitWorkerAndCloseOutput(waitGroup)
+}
+
+func (e *WindowParallelExec) waitWorkerAndCloseOutput(waitGroup *sync.WaitGroup) {
+	waitGroup.Wait()
+	close(e.outputCh)
+}
+
+// Next implements the Executor Next interface.
+func (e *WindowParallelExec) Next(ctx context.Context, req *chunk.Chunk) error {
+	req.Reset()
+	if !e.prepared {
+		e.prepare4ParallelExec(ctx)
+		e.prepared = true
+	}
+
+	failpoint.Inject("windowParallelError", func(val failpoint.Value) {
+		if val.(bool) {
+			failpoint.Return(errors.New("WindowParallelExec.Next error"))
+		}
+	})
+
+	if e.executed {
+		return nil
+	}
+	for {
+		result, ok := <-e.outputCh
+		if !ok {
+			e.executed = true
+			return nil
+		}
+		if result.err != nil {
+			return result.err
+		}
+		req.SwapColumns(result.chk)
+		result.chk.Reset()
+		result.giveBackCh <- result.chk
+		if req.NumRows() > 0 {
+			return nil
+		}
+	}
+}
+
+type windowDataFetcherExec struct {
+	baseExecutor
+	finishCh   <-chan struct{}
+	inputCh    <-chan *chunk.Chunk
+	giveBackCh chan<- *chunk.Chunk
+
+	executed bool
+}
+
+func (e *windowDataFetcherExec) Open(ctx context.Context) error {
+	if err := e.baseExecutor.Open(ctx); err != nil {
+		return err
+	}
+	e.executed = false
+	return nil
+}
+
+func (e *windowDataFetcherExec) Close() error {
+	return errors.Trace(e.baseExecutor.Close())
+}
+
+func (e *windowDataFetcherExec) Next(ctx context.Context, req *chunk.Chunk) error {
+	req.Reset()
+	if e.executed {
+		return nil
+	}
+	select {
+	case <-e.finishCh:
+		e.executed = true
+		return nil
+	case result, ok := <-e.inputCh:
+		if !ok || result.NumRows() == 0 {
+			e.executed = true
+			return nil
+		}
+		req.SwapColumns(result)
+		result.Reset()
+		e.giveBackCh <- result
+		return nil
+	}
+}
+
+func (w *windowParallelWorker) run(ctx context.Context, waitGroup *sync.WaitGroup) {
+	defer func() {
+		if r := recover(); r != nil {
+			recoveryWindowParallelExec(w.outputCh, r)
+		}
+		waitGroup.Done()
+	}()
+
+	for {
+		select {
+		case <-w.finishCh:
+			return
+		case chk, ok := <-w.outputHolderCh:
+			if !ok {
```

Comment:
We can leave out this judgment since it will catched by `<-w.finishCh`

Reply:
En...I think it would be more robust with this judgment, even though it is not necessary now.
What do you think ?

Reply:
ditto.